stan/math/prim/prob/double_exponential_lpdf.hpp

```diff
@@ -53,52 +56,57 @@ return_type_t<T_y, T_loc, T_scale> double_exponential_lpdf(
+  const auto& mu_col = as_column_vector_or_scalar(mu_ref);
+  const auto& sigma_col = as_column_vector_or_scalar(sigma_ref);
 
-  VectorBuilder<true, T_partials_return, T_scale> inv_sigma(size_sigma);
-  VectorBuilder<include_summand<propto, T_scale>::value, T_partials_return,
-                T_scale>
-      log_sigma(size_sigma);
-  for (size_t i = 0; i < size_sigma; i++) {
-    const T_partials_return sigma_dbl = value_of(sigma_vec[i]);
-    inv_sigma[i] = inv(sigma_dbl);
-    if (include_summand<propto, T_scale>::value) {
-      log_sigma[i] = log(sigma_dbl);
-    }
-  }
+  const auto& y_arr = as_array_or_scalar(y_col);
+  const auto& mu_arr = as_array_or_scalar(mu_col);
+  const auto& sigma_arr = as_array_or_scalar(sigma_col);
 
-  for (size_t n = 0; n < N; n++) {
-    const T_partials_return y_dbl = value_of(y_vec[n]);
-    const T_partials_return mu_dbl = value_of(mu_vec[n]);
-    const T_partials_return y_m_mu = y_dbl - mu_dbl;
-    const T_partials_return scaled_diff = fabs(y_m_mu) * inv_sigma[n];
+  ref_type_t<decltype(value_of(y_arr))> y_val = value_of(y_arr);
+  ref_type_t<decltype(value_of(mu_arr))> mu_val = value_of(mu_arr);
+  ref_type_t<decltype(value_of(sigma_arr))> sigma_val = value_of(sigma_arr);
 
-    if (include_summand<propto>::value) {
-      logp -= LOG_TWO;
-    }
-    if (include_summand<propto, T_scale>::value) {
-      logp -= log_sigma[n];
-    }
-    logp -= scaled_diff;
+  check_finite(function, "Random variable", y_val);
+  check_finite(function, "Location parameter", mu_val);
+  check_positive_finite(function, "Scale parameter", sigma_val);
+
+  const auto& inv_sigma = to_ref(inv(sigma_val));
+  const auto& y_m_mu
+      = to_ref_if<!is_constant_all<T_y, T_loc>::value>(y_val - mu_val);
+  const auto& abs_diff_y_mu = fabs(y_m_mu);
+  const auto& scaled_diff
+      = to_ref_if<!is_constant_all<T_scale>::value>(abs_diff_y_mu * inv_sigma);
 
-    T_partials_return rep_deriv
-        = is_constant_all<T_y, T_loc>::value ? 0 : sign(y_m_mu) * inv_sigma[n];
+  size_t N = max_size(y, mu, sigma);
+  if (include_summand<propto>::value) {
+    logp -= N * LOG_TWO;
+  }
+  if (include_summand<propto, T_scale>::value) {
+    logp -= sum(log(sigma_val)) * N / size(sigma);
+  }
+  logp -= sum(scaled_diff);
+
+  if (!is_constant_all<T_y, T_loc>::value) {
+    const auto& diff_sign = sign(y_m_mu);
+    const auto& rep_deriv
+        = to_ref_if<(!is_constant_all<T_y>::value
+                     && !is_constant_all<T_loc>::value)>(diff_sign * inv_sigma);
     if (!is_constant_all<T_y>::value) {
-      ops_partials.edge1_.partials_[n] -= rep_deriv;
+      ops_partials.edge1_.partials_ = -rep_deriv;
     }
     if (!is_constant_all<T_loc>::value) {
-      ops_partials.edge2_.partials_[n] += rep_deriv;
-    }
-    if (!is_constant_all<T_scale>::value) {
-      ops_partials.edge3_.partials_[n] -= inv_sigma[n] * (1 - scaled_diff);
+      ops_partials.edge2_.partials_ = std::move(rep_deriv);
```

Comment:
I'm gonna add comments on the moves (regarding the comment in exp mod normal). If it turns out we keep them, ignore these comments.